packages/caravan-wallets/src/bitbox.ts

```diff
@@ -0,0 +1,551 @@
+    messages.push({
+      state: PENDING,
+      level: INFO,
+      text: "Please plug in your BitBox.",
+      code: "device.setup",
+    });
+    messages.push({
+      state: ACTIVE,
+      level: INFO,
+      text: "Communicating with BitBox...",
+      code: "device.active",
+    });
+    return messages;
+  }
+
+  showPairingCodePopup(pairingCode: string): (() => void) | null {
+    if (this.showPairingCode) {
+      return this.showPairingCode(pairingCode);
+    }
+    const htmlContent = `
+<!doctype html>
+<html>
+  <head>
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta charset="utf-8">
+    <title>BitBox02 pairing</title>
+  </head>
+  <body>
+    <h1 style="text-align:center;">BitBox02 pairing code</h1>
+    <div id="code" style="white-space:pre;font-family:monospace,monospace;text-align:center;font-size:large;">${pairingCode}</div>
+  </body>
+</html>
+`;
+    const popup = window.open(
+      '',
+      'popupWindow',
+      'width=400,height=300',
+    );
+    if (popup) {
+      popup.document.write(htmlContent);
+      popup.document.close();
+      return () => {
+        popup.close();
+      };
+    }
+    return null;
+  }
+
+  async withDevice<T>(f: (device: PairedBitBox) => Promise<T>): Promise<T> {
+    const bitbox = await import('bitbox-api');
+
+    let hidePairingCode: (() => void) | null = null;
+    try {
+      const unpaired = await bitbox.bitbox02ConnectAuto(() => {
+        if (hidePairingCode) {
+          hidePairingCode();
+        }
+      })
+      const pairing = await unpaired.unlockAndPair()
+      const pairingCode = pairing.getPairingCode()
+      if (pairingCode) {
+        hidePairingCode = this.showPairingCodePopup(pairingCode);
+        // TODO: display pairing code to the user.
+        console.log('Pairing code:', pairingCode);
+      }
+      const pairedBitBox = await pairing.waitConfirm()
+      if (hidePairingCode) {
+        hidePairingCode();
+      }
+      try {
+        return await f(pairedBitBox)
+      } finally {
+        pairedBitBox.close()
+      }
+    } catch (err) {
+      const typedErr = bitbox.ensureError(err)
+      const isErrorUnknown = typedErr.code === 'unknown-js'
+      const errorMessage = isErrorUnknown ? typedErr.err! : typedErr.message
+      throw new Error(errorMessage);
+    } finally {
+      if (hidePairingCode) {
+        hidePairingCode();
+      }
+    }
+  }
+
+  async maybeRegisterMultisig(pairedBitBox: PairedBitBox, walletConfig: MultisigWalletConfig): Promise<{ scriptConfig: BtcScriptConfig, keypathAccount: string; }> {
+    const { scriptConfig, keypathAccount } = await convertMultisig(pairedBitBox, walletConfig);
+    const isRegistered = await pairedBitBox.btcIsScriptConfigRegistered(
+      convertNetwork(walletConfig.network),
+      scriptConfig,
+      keypathAccount,
+    );
+    // No name means the user inputs it on the device.
+    // eslint-disable-next-line no-undefined
+    const name = undefined;
```

Comment:
There is a name that comes in from the wallet config actually so maybe it should just use that (same is done for ledger and coldcard)

Reply:
There might be trouble as the BitBox02 has some requirements on the name - max 30 chars, only printalble ASCII chars (including space), etc. If entered on the device it cannot fail.